framework/include/base/DependencyResolverInterface.h

```diff
@@ -15,49 +15,144 @@
 #ifndef DEPENDENCYRESOLVERINTERFACE_H
 #define DEPENDENCYRESOLVERINTERFACE_H
 
+/// STL includes
 #include <string>
 #include <set>
 #include <iostream>
 #include <algorithm>
 
+// MOOSE incluces
+#include "InputParameters.h"
 #include "DependencyResolver.h"
+#include "MooseUtils.h"
 
+class DependencyResolverInterface;
+
+template<>
+InputParameters validParams<DependencyResolverInterface>();
+
+/**
+ * A class for sorting objects based on inter dependencies
+ */
 class DependencyResolverInterface
 {
 public:
-  DependencyResolverInterface() {}
+  DependencyResolverInterface(){}
 
+  /**
+   * Return a list of items being requested by an object
+   */
   virtual const std::set<std::string> & getRequestedItems() = 0;
+
+  /**
+   * Return a list of items that are owned by an object
+   */
   virtual const std::set<std::string> & getSuppliedItems() = 0;
 
+  /**
+   * Return a list of items that are iterable
+   *
+   * Iterable items are allowed to have cyclic dependencies, see the Material object
+   * and test/tests/materials/recompute/recompute.i to see how this is utilized.
+   */
+  virtual const std::set<std::string> & getIterativeItems(){ return _empty_set; }
```

Comment:
I guess I need to think about this interface a little bit more. I'm not sure I like the name `IterativeItems` but I don't have a better one at the moment. I terms of shared pointers, these are still dependent items but they are "weakly" dependent. Your extra logic is all in place to avoid inserting items into the `DependencyResolver` that would cause a cycle. Can we say with certainly though what the correct order should be if we don't insert those items at all before calling sort? We may be able to handle this case right at the `DependencyResolver` level.